### ISSUE TITLE
Extend coverage of sshnoports e2e tests

### ISSUE THREAD

**Issue by gkc** [atsign-foundation/noports#446]
### Is your feature request related to a problem? Please describe.

Currently we are testing reverse ssh with the daemon using /usr/bin/ssh to make the initial ssh tunnel, across various combinations of sshnp (client) and sshnpd (daemon). Currently untested is to have the daemon use the pure-dart SSHClient to make the initial ssh tunnel.

We now have support for forward ssh, with the client making the initial ssh tunnel - this is currently uncovered by e2e tests. For forward ssh we also have two options - using /usr/bin/ssh and using pure-dart SSHClient.

So while we currently cover the interoperability dimensions, we only do so with one of the four possible combinations of forward - reverse and pure-dart - /usr/bin/ssh

### Describe the solution you'd like

- (Must have) Add testing of forward ssh (initially for local-to-local only)
  - with the client using `--ssh-client pure-dart`
  - with the client using `--ssh-client /usr/bin/ssh`
- (Nice to have) Add testing of reverse ssh
  - with the daemon using `--ssh-client pure-dart`

Note: These tests can only be added either (1) as part of #443 or (2) after #443 has been merged to trunk

**Comment by gkc** [comment]
@JeremyTubongbanua @XavierChanth I need some assistance in figuring out the best way to extend / evolve what we have now to add the coverage described above

**Comment by JeremyTubongbanua** [comment]
@XavierChanth @gkc I was thinking of something like this:

- [ ] Forward SSH (6 tests)
    - pure dart
    - [ ] local (—ssh-client=pure-dart) vs. local
    - [ ] local (—ssh-client=pure-dart) vs. trunk
    - [ ] trunk (—ssh-client=pure-dart) vs. local
    - usr/bin/ssh:
    - [ ] local (—ssh-client=/usr/bin/ssh) vs. local  
    - [ ] local (—ssh-client=/usr/bin/ssh) vs. trunk
    - [ ] trunk (—ssh-client=/usr/bin/ssh) vs. local    
- [ ] Reverse SSH (3 test)
    - [ ] local vs. local (--ssh-client pure-dart)
    - [ ] local vs. trunk (--ssh-client pure-dart)
    - [ ] trunk vs. local (--ssh-client pure-dart)  


@gkc For the reverse ssh, the daemon will have `--ssh-client=pure-dart` and the client **must** have no `--ssh-client` flags, is that right?

**Comment by gkc** [comment]
> @XavierChanth @gkc I was thinking of something like this:
> 
> 
> 
> - [ ] Forward SSH (6 tests)
> 
>     - pure dart
> 
>     - [ ] local (—ssh-client=pure-dart) vs. local
> 
>     - [ ] local (—ssh-client=pure-dart) vs. trunk
> 
>     - [ ] trunk (—ssh-client=pure-dart) vs. local
> 
>     - usr/bin/ssh:
> 
>     - [ ] local (—ssh-client=/usr/bin/ssh) vs. local  
> 
>     - [ ] local (—ssh-client=/usr/bin/ssh) vs. trunk
> 
>     - [ ] trunk (—ssh-client=/usr/bin/ssh) vs. local    
> 
> - [ ] Reverse SSH (3 test)
> 
>     - [ ] local vs. local (--ssh-client pure-dart)
> 
>     - [ ] local vs. trunk (--ssh-client pure-dart)
> 
>     - [ ] trunk vs. local (--ssh-client pure-dart)  
> 
> 
> 
> 


Sounds good yep that's what I had in mind - what's the best way to do it?


> 
> @gkc For the reverse ssh, the daemon will have `--ssh-client=pure-dart` and the client **must** have no `--ssh-client` flags, is that right?

Re daemon that's correct. Re client - for reverse ssh the client doesn't  execute an ssh so the flag currently doesn't matter. Likewise for direct ssh the daemon doesn't execute an ssh, so the flag doesn't matter to it.

**Comment by JeremyTubongbanua** [comment]
We could add more entrypoints like this:

![image](https://github.com/atsign-foundation/sshnoports/assets/79019866/6566cf56-8463-4d9f-9941-cd2c81cff5e4)

Then write additional jobs.

![image](https://github.com/atsign-foundation/sshnoports/assets/79019866/c7df8be2-979a-472b-9392-c22b13f3d03a)

We could add 2 jobs (forward ssh /usr/bin/ssh, forward ssh pure-dart), each with a matrix of 3 items (local-local, local-trunk, trunk-local).

Our current `e2e_test` doesn't have any flags, so that would be using reverse ssh by default, if I am understanding correctly. So we would rename our current `e2e_test` to `e2e_reverse_ssh_test`

**Comment by gkc** [comment]
Lots of progress during PR78 after much effort, almost done. Estimate another 3SP to finish off during PR79

**Comment by gkc** [comment]
Little progress in PR79; 2 SP to complete in PR80

**Comment by gkc** [comment]
No time spent during PR81, will prioritize to complete and close in PR82

**Comment by gkc** [comment]
Completed by #888 burned 8 SP this sprint 🫢

New tickets #890 and #891
